docs/fassets/developer-guides/12-fassets-autoredeem.mdx

```diff
@@ -0,0 +1,649 @@
+
+```bash
+# Step 2: Auto-redeem back to native XRP
+yarn hardhat run scripts/fassets/autoRedeemHyperEVM.ts --network hyperliquidTestnet
+```
+
+---
+
+## Overview
+
+This system enables developers to bridge FAssets (specifically FXRP) between Flare's Coston2 testnet and Hyperliquid EVM Testnet using LayerZero's cross-chain messaging protocol.
+It supports two key functionalities:
+
+1. **Bridging FXRP from Coston2 to Hyperliquid EVM** (`bridgeToHyperEVM.ts`) - Transfer wrapped XRP tokens to Hyperliquid for trading or DeFi use
+2. **Auto-Redeem from Hyperliquid to Underlying Asset** (`autoRedeemHyperEVM.ts`) - Send FXRP from Hyperliquid back to Coston2 and automatically redeem it for native XRP
+
+**These are sequential operations:** You must complete Step 1 (bridging TO Hyperliquid) before you can execute Step 2 (auto-redeeming FROM Hyperliquid).
+
+---
+
+## FAssetRedeemComposer Contract
+
+### What It Is
+
+`FAssetRedeemComposer.sol` is a LayerZero Composer contract that automatically redeems FAssets to their underlying assets when tokens arrive on Coston2 via LayerZero's compose message feature.
+
+### How It Works
+
+The composer implements the `IOAppComposer` interface and processes LayerZero compose messages:
+
+1. **Receives Compose Message**: LayerZero endpoint calls `lzCompose()` with the incoming OFT transfer and compose data
+2. **Extracts Parameters**: Decodes the compose message to get:
+   - Amount to redeem (in lots)
+   - Underlying address (XRP address)
+   - Redeemer address (EVM address)
+3. **Gets Asset Manager**: Retrieves the FAsset AssetManager from Flare's ContractRegistry
+4. **Calculates Lots**: Determines how many lots can be redeemed based on received balance
+5. **Approves Tokens**: Grants AssetManager permission to spend the FAsset tokens
+6. **Executes Redemption**: Calls `assetManager.redeem()` to burn FAssets and release underlying XRP
+
+### Key Features
+
+- **Automatic Execution**: No manual redemption step required
+- **Reentrancy Protected**: Uses OpenZeppelin's `ReentrancyGuard`
+- **Owner Controls**: Allows owner to recover stuck tokens or native currency
+- **Event Emission**: Emits `RedemptionTriggered` event for tracking
+
+### Error Handling
+
+- `OnlyEndpoint()`: Reverts if caller is not the LayerZero endpoint
+- `InsufficientBalance()`: Reverts if no tokens were received
+- `AmountTooSmall()`: Reverts if received amount is less than 1 lot
+
+<details>
+<summary>View `FAssetRedeemComposer.sol` source code</summary>
+
+<CodeBlock language="solidity" title="contracts/FAssetRedeemComposer.sol">
+  {FassetRedeemComposer}
+</CodeBlock>
+
+</details>
+
+### Deployment
+
+Deploy the composer contract to Coston2:
+
+```bash
+# Set deployment configuration
+export DEPLOYER_PRIVATE_KEY=your_private_key
+
+# Deploy via Hardhat deploy scripts
+npx hardhat deploy --network coston2 --tags FAssetRedeemComposer
+
+# Save the deployed address to your .env
+export COSTON2_COMPOSER=0x...
+```
+
+---
```

Comment:
I think can remove this line and have a paragraph what is this bridging to Hyperliquid and then start the guide. Basically to give context for users who jump to this section.